### PR TITLE
ttl: support to split TTL tasks for a table without binary primary key

### LINKED ISSUE #55660: ttl: support split TTL tasks for table with non-binary primary key
## Enhancement

see description in PR #55660

### PR DESCRIPTION
**Pull request by lcwangchao**
### What problem does this PR solve?

Issue Number: close #55660

Problem Summary:

When running a TTL job, the job will be split into several tasks to delete rows of concurrency. However, if a table's primary key is a string type without a binary flag, the task cannot be split. For example:

```
CREATE TABLE t(id varchar(32) CHARACTER SET UTF8MB4, t timestamp) TTL=`t` + INTERVAL 1 HOUR;
```
When table t starts a TTL job, it only has one task even if it is a large table with many regions.

The reason is that we split tasks according to the region edges in the current implementation. However, the region is a conception in TiKV which will not ensure it will have a legal encoding with the table record. If use the raw bytes decoded from the region edge as the SQL condition, it will report an error.

For example, for the above table `t`, the below SQL will report an error:

```
> select * from t where id < x'ff';
(1105, "Cannot convert string '\\xFF' from binary to utf8mb4")
```

To avoid this, we did not do split works for this type of table.

But in practice, using a string column as the primary key is very normal, such as to store uuid, phone number, etc... Though we can use binary charset to walk around it to make TTL split tasks, it is some what strange and not easy to use.

### What changed and how does it work?

We support to split TTL tasks for tables with string columns without binary flag.  In this PR, we only parse the visible ASCII prefix of each region edge bytes. For example, if a table has some regions:

```
region1: recordPrefix -> recordPrefix + encodedBytes('abc\xffde')
region2: recordPrefix + encodedBytes('abc\xffde') -> recordPrefix + encodedBytes('abd\xffxyz') 
region2: encodedBytes('abc\xde') -> end
```

3 tasks in TTL job will be created with primary ranges:

- `id < 'abc'`
- `id >= 'abc' AND id < 'abd'`
- `id > 'abd'`

Though it still has some limitations such as it still cannot work well when most keys does not contain ASCII prefix or share the same ASCII prefix, it can fit most scenes.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)

tested in a real cluster:

```
> show create table t1;
+-------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Table | Create Table                                                                                                                                                      |
+-------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| t1    | CREATE TABLE `t1` (                                                                                                                                               |
|       |   `id` varchar(64) NOT NULL,                                                                                                                                      |
|       |   `t` timestamp NULL DEFAULT NULL,                                                                                                                                |
|       |   PRIMARY KEY (`id`) /*T![clustered_index] CLUSTERED */                                                                                                           |
|       | ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin /*T![ttl] TTL=`t` + INTERVAL 0 DAY */ /*T![ttl] TTL_ENABLE='ON' */ /*T![ttl] TTL_JOB_INTERVAL='1m' */ |
+-------+-------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set
Time: 0.019s
```

insert several rows, we can see the TTL tasks are splitter to several tasks:

```
> select scan_id, state, status from mysql.tidb_ttl_task;
+---------+-------------------------------------------------------------------------------+----------+
| scan_id | state                                                                         | status   |
+---------+-------------------------------------------------------------------------------+----------+
| 0       | {"total_rows":821229,"success_rows":821229,"error_rows":0,"scan_task_err":""} | finished |
| 1       | {"total_rows":524288,"success_rows":524288,"error_rows":0,"scan_task_err":""} | finished |
| 2       | {"total_rows":296942,"success_rows":296942,"error_rows":0,"scan_task_err":""} | finished |
| 3       | {"total_rows":296942,"success_rows":296942,"error_rows":0,"scan_task_err":""} | finished |
| 4       | {"total_rows":454692,"success_rows":454692,"error_rows":0,"scan_task_err":""} | finished |
| 5       | {"total_rows":296942,"success_rows":296942,"error_rows":0,"scan_task_err":""} | finished |
| 6       | {"total_rows":296942,"success_rows":296942,"error_rows":0,"scan_task_err":""} | finished |
| 7       | {"total_rows":296942,"success_rows":296942,"error_rows":0,"scan_task_err":""} | finished |
| 8       | {"total_rows":296942,"success_rows":296942,"error_rows":0,"scan_task_err":""} | finished |
| 9       | {"total_rows":296942,"success_rows":296942,"error_rows":0,"scan_task_err":""} | finished |
| 10      | {"total_rows":296942,"success_rows":296942,"error_rows":0,"scan_task_err":""} | finished |
| 11      | {"total_rows":315500,"success_rows":315500,"error_rows":0,"scan_task_err":""} | finished |
| 12      | {"total_rows":296942,"success_rows":296942,"error_rows":0,"scan_task_err":""} | finished |
| 13      | {"total_rows":296942,"success_rows":296942,"error_rows":0,"scan_task_err":""} | finished |
| 14      | {"total_rows":296942,"success_rows":296942,"error_rows":0,"scan_task_err":""} | finished |
| 15      | {"total_rows":296942,"success_rows":296942,"error_rows":0,"scan_task_err":""} | finished |
| 16      | {"total_rows":296942,"success_rows":296942,"error_rows":0,"scan_task_err":""} | finished |
| 17      | {"total_rows":296942,"success_rows":296942,"error_rows":0,"scan_task_err":""} | finished |
| 18      | {"total_rows":296942,"success_rows":296942,"error_rows":0,"scan_task_err":""} | finished |
| 19      | {"total_rows":296942,"success_rows":296942,"error_rows":0,"scan_task_err":""} | finished |
| 20      | {"total_rows":296942,"success_rows":296942,"error_rows":0,"scan_task_err":""} | finished |
| 21      | {"total_rows":296942,"success_rows":296942,"error_rows":0,"scan_task_err":""} | finished |
| 22      | {"total_rows":296942,"success_rows":296942,"error_rows":0,"scan_task_err":""} | finished |
| 23      | {"total_rows":296942,"success_rows":296942,"error_rows":0,"scan_task_err":""} | finished |
| 24      | {"total_rows":296942,"success_rows":296942,"error_rows":0,"scan_task_err":""} | finished |
| 25      | {"total_rows":37117,"success_rows":37117,"error_rows":0,"scan_task_err":""}   | finished |
+---------+-------------------------------------------------------------------------------+----------+
```

- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
ttll: support to split job to several tasks for `utf8` and `utf8mb4` keys.
```

### CODE REVIEW

File: pkg/ttl/cache/table.go
Comment (YangKeao) [inline pkg/ttl/cache/table.go:296]: Should it only accept `utf8mb4_bin` (or other binary collations)?

For `utf8mb4_unicode_ci` (or other UCA collations), there is no direct connection between the character and the weight. It's meaningless to get the ASCII prefix.

Comment (YangKeao) [reply pkg/ttl/cache/table.go:296]: For example, the string `a` has weight `0x0E33`.

Comment (lcwangchao) [reply pkg/ttl/cache/table.go]: Done, PTAL again

File: pkg/ttl/cache/table.go
Comment (bb7133) [inline pkg/ttl/cache/table.go:305]: I don't think collations matter here: they don't change the encoding of record keys.

**Review by bb7133 — APPROVED** [review]
LGTM

**Comment by lcwangchao** [comment]
> Do we need to update the doc, too?

Yes, I'll update the doc after this PR merged

**Review by YangKeao — APPROVED** [review]
LGTM

**Comment by YangKeao** [comment]
/hold

`utf8mb4_0900_bin` is also acceptable.

File: pkg/ttl/cache/table.go
Comment (YangKeao) [inline pkg/ttl/cache/table.go:303]: `utf8mb4_0900_bin` is also acceptable. I believe the only difference between `utf8mb4_0900_bin` and `utf8mb4_bin` is that it's `NO PAD`.

**Pull request merged**
